validator/sawtooth_validator/server/state_verifier.py

```diff
@@ -0,0 +1,287 @@
+    blockstore = BlockStore(block_db)
+
+    # Check if we should do state verification
+    start = binary_search_for_state_root(blockstore, state_view_factory)
+    if start is None:
+        LOGGER.info(
+            "Skipping state verification: chain head's state root is present")
+        return
+
+    LOGGER.info(
+        "Recomputing missing state from block %s with %s scheduler",
+        start, scheduler_type)
+
+    component_thread_pool = InstrumentedThreadPoolExecutor(
+        max_workers=10,
+        name='Component')
+
+    component_dispatcher = Dispatcher()
+    component_service = Interconnect(
+        bind_component,
+        component_dispatcher,
+        secured=False,
+        heartbeat=False,
+        max_incoming_connections=20,
+        monitor=True,
+        max_future_callback_workers=10)
+
+    context_manager = ContextManager(global_state_db)
+
+    transaction_executor = TransactionExecutor(
+        service=component_service,
+        context_manager=context_manager,
+        settings_view_factory=SettingsViewFactory(state_view_factory),
+        scheduler_type=scheduler_type,
+        invalid_observers=[])
+
+    component_service.set_check_connections(
+        transaction_executor.check_connections)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_RECEIPT_ADD_DATA_REQUEST,
+        tp_state_handlers.TpReceiptAddDataHandler(context_manager),
+        component_thread_pool)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_EVENT_ADD_REQUEST,
+        tp_state_handlers.TpEventAddHandler(context_manager),
+        component_thread_pool)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_STATE_DELETE_REQUEST,
+        tp_state_handlers.TpStateDeleteHandler(context_manager),
+        component_thread_pool)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_STATE_GET_REQUEST,
+        tp_state_handlers.TpStateGetHandler(context_manager),
+        component_thread_pool)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_STATE_SET_REQUEST,
+        tp_state_handlers.TpStateSetHandler(context_manager),
+        component_thread_pool)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_REGISTER_REQUEST,
+        processor_handlers.ProcessorRegisterHandler(
+            transaction_executor.processors),
+        component_thread_pool)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_UNREGISTER_REQUEST,
+        processor_handlers.ProcessorUnRegisterHandler(
+            transaction_executor.processors),
+        component_thread_pool)
+
+    component_dispatcher.start()
+    component_service.start()
+
+    process_blocks(
+        blocks=blockstore.get_block_iter(start_block=start, reverse=False),
+        transaction_executor=transaction_executor,
+        context_manager=context_manager,
+        state_view_factory=state_view_factory)
+
```

Comment:
the ContextManager and the Executor are both active components that need to be shut down. I think they both have a stop method.

Reply:
fixed